redfish/collector.go

```diff
@@ -0,0 +1,241 @@
+}
+
+// CollectorConfig is a set of configurations for Collector.
+type CollectorConfig struct {
+	AddressConfig *config.AddressConfig
+	Port          string
+	UserConfig    *config.UserConfig
+	Rule          []byte
+}
+
+// NewCollector returns a new instance of Collector.
+func NewCollector(cc *CollectorConfig) (*Collector, error) {
+	cache := &cache{dataMap: make(dataMap)}
+	client, err := newClient(cc, cache)
+	if err != nil {
+		return nil, err
+	}
+
+	var rules []convertRule
+	err = yaml.Unmarshal(cc.Rule, &rules)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, rule := range rules {
+		if err := rule.Validate(); err != nil {
+			return nil, err
+		}
+	}
+
+	return &Collector{rules: rules, client: client, cache: cache}, nil
+}
+
+// Describe does nothing for now.
+func (c Collector) Describe(ch chan<- *prometheus.Desc) {
+}
+
+// Collect sends metrics collected from BMC via Redfish.
+func (c Collector) Collect(ch chan<- prometheus.Metric) {
+	dataMap := c.cache.get()
+
+	for _, rule := range c.rules {
+		for path, parsedJSON := range dataMap {
+			matched, pathLabels := matchPath(rule.Path, path)
+			if !matched {
+				continue
+			}
+
+			for _, propertyRule := range rule.Rules {
+				matchedProperties := matchPointer(propertyRule.Pointer, parsedJSON, path)
+				for _, matched := range matchedProperties {
+					value, err := propertyRule.Converter(matched.property)
+					if err != nil {
+						log.Warn("failed to interpret Redfish data as metric", map[string]interface{}{
+							"path":      path,
+							"pointer":   propertyRule.Pointer,
+							"name":      propertyRule.Name,
+							"value":     matched.property,
+							log.FnError: err,
+						})
+						continue
+					}
+					labels := make(prometheus.Labels)
+					for k, v := range pathLabels {
+						labels[k] = v
+					}
+					for k, v := range matched.indexes {
+						labels[k] = strconv.Itoa(v)
+					}
+					desc := prometheus.NewDesc(
+						prometheus.BuildFQName(namespace, "", propertyRule.Name),
+						propertyRule.Description, nil, labels)
+					ch <- prometheus.MustNewConstMetric(desc, prometheus.GaugeValue, value)
+				}
+			}
+		}
+	}
+}
+
+// Update collects metrics from BMCs via Redfish.
+func (c *Collector) Update(ctx context.Context, rootPath string) {
+	c.client.update(ctx, rootPath)
+}
+
+func matchPath(rulePath, path string) (bool, prometheus.Labels) {
+	ruleElements := strings.Split(rulePath, "/")
+	pathElements := strings.Split(path, "/")
+
+	if len(ruleElements) != len(pathElements) {
+		return false, nil
+	}
+
+	labels := make(prometheus.Labels)
+	for i := 0; i < len(ruleElements); i++ {
+		ln := len(ruleElements[i])
+		if ln >= 2 && ruleElements[i][0] == '{' && ruleElements[i][ln-1] == '}' {
+			labelName := ruleElements[i][1 : ln-1]
+			labels[labelName] = pathElements[i]
+		} else if ruleElements[i] != pathElements[i] {
+			return false, nil
+		}
+	}
+
+	return true, labels
+}
+
+type dataMap map[string]*gabs.Container
+
+type cache struct {
```

Comment:
For this type of atomics, use `atomic.Value`.  It scales better for in mostly read cases.